### PR TITLE
libserver Makefile changes

### PR DESCRIPTION
**Pull request by jvaltin**
This is a minimally invasive set of targets that should document the libserver build process for our three major supported platforms.

### CODE REVIEW

File: frontends/qt/server/Makefile.linux [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+BUILDMODE=c-shared
+GOARCH=amd64
+GOOS=linux
+
+linux:
+	CGO_CFLAGS="-g ${GOLNXSECFLAGS} ${CFLAGS}" \
+	CGO_LDFLAGS="${GOLNXLDFLAGS} ${LFLAGS}" \
+	GOARCH=${GOARCH} CGO_ENABLED=${CGO} GOOS=${GOOS} \
+	go build -x -buildmode=${BUILDMODE} -ldflags \
+		-extldflags=${GOLNXLDFLAGS} -o ${LIBNAME}.so
```
[the remark this answers is not in the document]
Comment (x1ddos) [reply frontends/qt/server/Makefile.linux]: Then in the final app output you'd have something like:

bitbox-bin: $(LIBNAME) main.c or whatever
    ...build commands...

[the remark this answers is not in the document]
Comment (jvaltin) [reply frontends/qt/server/Makefile.linux]: I would prefer that we use autotools correctly over shell scripts.

**Comment by x1ddos** [comment]
Autotools is a separate topic.

**Comment by jvaltin** [comment]
> Autotools is a separate topic.

I would prefer to rewrite the build system in the manner you suggest. That set of changes would be invasive and possibly confusing to those who aren't heavy automake users. It also seems like an orthogonal set of changes to the set of changes that I proposed here.

Should such a rewrite occur before my changeset relating to setting golang build environment variables will be considered, or after review ready for merge? Are you asking that I rework only the proposed Makefile targets to be less like the current style, and more as you suggest?

**Comment by x1ddos** [comment]
> That set of changes would be invasive and possibly confusing to those who aren't heavy automake 

I very much disagree. I consider myself a newbie in automake/autoconf tools and never use them at all. Yes, this has never prohibited me from leveraging all the nice features and power of make in many projects.

> Are you asking that I rework only the proposed Makefile targets to be less like the current style, and more as you suggest?

Sorry, this was a general suggestion. Might as well happen in a separate CL. Maybe even better. Never mind :)

**Comment by jvaltin** [comment]
> I very much disagree. I consider myself a newbie in automake/autoconf tools and never use them at all. Yes, this has never prohibited me from leveraging all the nice features and power of make in many projects.

I offered to make a similar set of changes and @benma did not seem to want sweeping changes to the style or substance of these Makefiles. I might have misunderstood.

> Sorry, this was a general suggestion. Might as well happen in a separate CL. Maybe even better. Never mind :)

OK. After landing these changes, I would be interested in helping with those changes.

**Comment by benma** [comment]
@jvaltin remembers correctly. I don't think it is worthwhile rewriting the whole build system, seeing that the project compiles very quickly, has barely any dependency chain, and autotools is a nightmare to maintain for people not familiar with it like myself. 

Let's not blow the task of adding some hardening out of proportion.

Let's review, test and merge this PR first before anything else.

**Comment by jvaltin** [comment]
I've added updates to the BitBox.pro file which allow me to build the `qt-linux` target from the top most Makefile. The results show the expected security improvements:
```
RELRO           STACK CANARY      NX            PIE             RPATH      RUNPATH	Symbols		FORTIFY	Fortified	Fortifiable  FILE
Full RELRO      Canary found      NX enabled    PIE enabled     No RPATH   No RUNPATH   No Symbols       Yes	5		12		frontends/qt/server/libserver.so
Full RELRO      Canary found      NX enabled    PIE enabled     No RPATH   RUNPATH      268 Symbols	Yes	0		2		frontends/qt/build/linux-tmp/BitBox
```

This branch should be tested on a Mac OS X machine and on a Windows machine.

**Comment by jvaltin** [comment]
> Error on linux:

The error seems to be that the libc6-dev package contains object code that wasn't compiled with -fPIC. Those either need to be rebuilt or a modern OS needs to be used. This is regarding Ubuntu 14.04 and Ubuntu gained -fPIC support widely in 16.10. Switching to a modern OS seems the most reasonable choice.

> Error on macOS

That's surprising as `GODARWINLDFLAGS=${LASLR},${LSTRIP},${LEXPORT}` does not appear to include `-z`.

**Comment by jvaltin** [comment]
The current set of revisions under consideration should now build on @benma's Windows development machine, a freshly installed MacOS machine (see #27), and in the GNU/Linux docker environment.

The Windows development machine appears to not have `libssp` and it does not appear to utilize  `QMAKE_CXXFLAGS` or `QMAKE_LFLAGS` as set in `BitBox.pro`.

**Comment by jvaltin** [comment]
I've found that the issue with building inside of dockerdev can be solved by setting $CC to clang - this requires the Docker image to have clang installed. This is probably related to [go issue 15038](https://github.com/golang/go/issues/15038#issuecomment-204207755) and we need to update the compiler/linker. Since this is for a small shared library and we use different versions, and even different compilers already, I think using clang to resolve this as a special measure is ok.

**Comment by jvaltin** [comment]
Adding clang requires a Docker image update. I've verified that the current branch results in a working Docker image by rebuilding with `make dockerinit` and subsequently confirmed that the `qt-linux` target works inside of the `dockerdev` environment. The resulting Debian package works and the BitBox program runs and behaves as expected on a Debian stable system.

**Comment by benma** [comment]
:tada: :dancing_men: thank you!

**Pull request merged**